SSA/Projects/CIRCT/comb/comb.lean

```diff
@@ -163,77 +163,84 @@ open TyDenote (toType)
 -- arg type CONF
 @[simp, reducible]
 def Op.sig : Op  → List Ty
-  | .add w => [Ty.bv w, Ty.bv w]
-  | .and w => [Ty.bv w, Ty.bv w]
-  | .concat w₁ w₂ => [Ty.bv w₁, Ty.bv w₂]
+  | .add w n => List.replicate (n + 1) (Ty.bv w)
+  | .and w n => List.replicate (n + 1) (Ty.bv w)
+  | .concat w  => w.map (fun w => Ty.bv w)
   | .divs w => [Ty.bv w, Ty.bv w]
   | .divu w => [Ty.bv w, Ty.bv w]
   | .extract w n => [Ty.bv w, Ty.nat n]
   | .icmp w n => [Ty.bv w, Ty.bv w, Ty.nat n]
   | .mods w => [Ty.bv w, Ty.bv w]
   | .modu w => [Ty.bv w, Ty.bv w]
-  | .mul w => [Ty.bv w, Ty.bv w]
+  | .mul w n => List.replicate (n + 1) (Ty.bv w)
   | .mux w => [Ty.bv w, Ty.bv w, Ty.bool]
-  | .or w => [Ty.bv w, Ty.bv w]
+  | .or w n => List.replicate (n + 1) (Ty.bv w)
   | .parity w => [Ty.bv w]
-  | .replicate w n => [Ty.bv w, Ty.nat n]
+  | .replicate w _ => [Ty.bv w]
   | .shl w => [Ty.bv w, Ty.bv w]
   | .shrs w => [Ty.bv w, Ty.bv w]
   | .shru w => [Ty.bv w, Ty.bv w]
   | .sub w => [Ty.bv w, Ty.bv w]
-  | .xor w => [Ty.bv w, Ty.bv w]
+  | .xor w n => List.replicate (n + 1) (Ty.bv w)
 
 -- return type CONF
 @[simp, reducible]
 def Op.outTy : Op  → Ty
-  | .add w => Ty.bv w
-  | .and w => Ty.bv w
-  | .concat w₁ w₂ => Ty.bv (w₁ + w₂)
+  | .add w _ => Ty.bv w
+  | .and w _ => Ty.bv w
+  | .concat w => Ty.bv w.sum
   | .divs w => Ty.bv w
   | .divu w => Ty.bv w
   | .extract w n => Ty.bv (w - n)
   | .icmp _ _ => Ty.bool
   | .mods w => Ty.bv w
   | .modu w => Ty.bv w
-  | .mul w => Ty.bv w
+  | .mul w _ => Ty.bv w
   | .mux w => Ty.bv w
-  | .or w =>  Ty.bv w
+  | .or w _ =>  Ty.bv w
   | .parity _ => Ty.bool
   | .replicate w n =>  Ty.bv (w * n)
   | .shl w =>  Ty.bv w
   | .shrs w =>  Ty.bv w
   | .shru w =>  Ty.bv w
   | .sub w =>  Ty.bv w
-  | .xor w => Ty.bv w
+  | .xor w _ => Ty.bv w
 
 @[simp, reducible]
 def Op.signature : Op → Signature (Ty) :=
   fun o => {sig := Op.sig o, outTy := Op.outTy o, regSig := []}
 
 instance : DialectSignature Comb := ⟨Op.signature⟩
 
+def HVector.toList : HVector f (List.replicate n a) → List (f a) :=
+  sorry
+
+def Comb.variadicAdd (l : List (BitVec w)) : List (BitVec w) → BitVec w :=
+  List.foldr BitVec.add _ _
+  sorry
+
```

Comment:
Let me know if you need help filling in these sorries! Especially the first one might be a bit tricky